3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java

```diff
@@ -1,21 +1,106 @@
+import org.springframework.cglib.proxy.Callback;
+import org.springframework.cglib.proxy.Enhancer;
+import org.springframework.cglib.proxy.MethodInterceptor;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
 
 /**
  * This is processor class implements {@link BeanPostProcessor}, looks for a beans where method parameters are marked with
  * {@link Trimmed} annotation, creates proxy of them, overrides methods and trims all {@link String} arguments marked with
  * {@link Trimmed}. For example if there is a string " Java   " as an input parameter it has to be automatically trimmed to "Java"
  * if parameter is marked with {@link Trimmed} annotation.
  * <p>
- *
+ * <p>
  * Note! This bean is not marked as a {@link Component} to avoid automatic scanning, instead it should be created in
  * {@link StringTrimmingConfiguration} class which can be imported to a {@link Configuration} class by annotation
  * {@link EnableStringTrimming}
  */
-public class TrimmedAnnotationBeanPostProcessor {
-//todo: Implement TrimmedAnnotationBeanPostProcessor according to javadoc
+public class TrimmedAnnotationBeanPostProcessor implements BeanPostProcessor {
+
+    private final Map<String, Object> beansToBeProxied = new HashMap<>();
+    private final Map<String, MethodInterceptor[]> beanInterceptors = new HashMap<>();
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        if (containsParametersAnnotatedWithTrimmed(bean)) {
+            beansToBeProxied.put(beanName, bean);
+            beanInterceptors.put(beanName, new MethodInterceptor[]{provideTrimmingInterceptor()});
+        }
+        return BeanPostProcessor.super.postProcessBeforeInitialization(bean, beanName);
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        Object toProxy = beansToBeProxied.get(beanName);
+        if (toProxy != null) {
+            Enhancer enhancer = new Enhancer();
+            enhancer.setSuperclass(toProxy.getClass());
+            setInterceptorsIfNeeded(enhancer, bean, beanName);
+            return enhancer.create();
+        }
+        return bean;
+    }
+
+    private MethodInterceptor provideTrimmingInterceptor() {
+        return (beanInstance, method, args, methodProxy) ->
+                methodProxy.invokeSuper(beanInstance, processParameters(method, args));
+    }
+
+    private Object[] processParameters(Method method, Object[] args) {
+        Parameter[] parameters = method.getParameters();
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i].isAnnotationPresent(Trimmed.class)) {
+                args[i] = trimMarkedString(args[i]);
+            }
+        }
+        return args;
+    }
+
+    private Object trimMarkedString(Object arg) {
```

Comment:
@Stanislav-Zabramnyi I would rather say 
```java
        if (arg != null && arg instanceof String string) {
            return string.trim();
        }
        return arg;
```

Reply:
condition `arg! = null` has already covered by `arg instanceof String`, if `arg is null` , `arg instanceof String` will just return `false`